Classes/Service/ImportService.php

```diff
@@ -0,0 +1,180 @@
+        ];
+
+        $importData = $this->loadDataFromJsonFile($importJob->getJsonFile());
+
+        foreach ($importData as $importPosition => $data) {
+            // cleanup array from not importable data lines
+            if (
+                !array_key_exists($importPosition, $importJob->getDatasetMapping())
+                || $importJob->getDatasetMapping()[$importPosition] !== '1'
+            ) {
+                unset($importData[$importPosition]);
+                continue;
+            }
+            $insertArray = [];
+            $update = false;
+            foreach ($importJob->getFieldMapping() as $position => $field) {
+                if ($field === 'uid') {
+                    if (!empty($data[$position])) {
+                        $update = true;
+                    }
+                } else {
+                    $insertArray[$field] = $data[$position];
+                }
+            }
+            if (!isset($insertArray['pid'])) {
+                $insertArray['pid'] = $importJob->getPid();
+            }
+            foreach ($importJob->getDefaultValues() as $fieldName => $defaultValue) {
+                $insertArray[$fieldName] = $defaultValue;
+            }
+
+            foreach ($importJob->getPasswordFields() as $passwordField) {
+                $insertArray[$passwordField] = md5(sha1(microtime()));
+            }
+
+            $identifier = $update ? $data['uid'] : StringUtility::getUniqueId('NEW');
+
+            $insertData[$importJob->getTable()][$identifier] = $insertArray;
+        }
+
+        /** @deprecated Function call will be removed in 6.0 */
+        $insertData = $this->callHook($insertData, $importJob->getTable());
+
+        /** @var ManipulateRelationsEvent $manipulateRelationsEvent */
+        $manipulateRelationsEvent = $this->eventDispatcher->dispatch(
+            new ManipulateRelationsEvent($importJob->getTable(), $insertData)
+        );
+
+        $importJob->setData($manipulateRelationsEvent->getInsertData());
+    }
+
+    public function writeImport(ImportJob $importJob): void
+    {
+        $cmd = [];
+        if ($importJob->isDeleteExisting()) {
+            $cmd = $this->prepareRecordDeletebyJob($importJob);
+        }
+        $dataHandler = GeneralUtility::makeInstance(DataHandler::class);
+        $dataHandler->start($importJob->getData(), $cmd);
+        $dataHandler->process_cmdmap();
+        $dataHandler->process_datamap();
+    }
+
+    /**
+     * @param array<string, array<int|string, array<string, mixed>>> $insertData
+     * @param string $table
+     * @return array<string, array<int|string, array<string, mixed>>>
+     * @deprecated Using Hooks is deprecated since 5.0. Use Events instead
+     */
+    private function callHook(array $insertData, string $table): array
+    {
+        if (count($GLOBALS['TYPO3_CONF_VARS']['SC_OPTIONS']['SUDHAUS7\\Xlsimport\\Controller\\XlsimportController']['Hooks'] ?? []) > 0) {
+            trigger_deprecation(
```

Comment:
Some additional remark here:

The TYPO3 core removes old hooks and introducing new event's in one go - instead of deprecating them. Therefore, extensions (in this case other extensions) can implement both variants, where the hook works for the old version and the event for the new version. No need for "version" checkes.

Therefore, I would suggest to consider to do this in a same manner.

Otherwise, if a follo-up extension implements boths, both would be called.

Reply:
changed. Thanks for the hint